WangscapeTest/TestCalculatorDither.cpp

```diff
@@ -0,0 +1,156 @@
+                    EXPECT_EQ(255, alphas[i]);
+                else
+                    EXPECT_EQ(0, alphas[i]);
+            }
+        }
+        if (expected_winners)
+        {
+            if (cd.power < 10)
+                EXPECT_EQ(expected_winners, results.size()) <<
+                "Unexpected number of winners. power = " << std::to_string(cd.power);
+            else
+                EXPECT_EQ(1, results.size()) <<
+                "Unexpected number of winners. power = " << std::to_string(cd.power);
+        }
+    }
+};
+
+TEST_F(TestCalculatorDither, TestCalculatorDitherAllEqual)
+{
+    for (double weight : weight_examples)
+    {
+        tilegen::alpha::CalculatorBase::Weights weights{weight, weight, weight, weight};
+        for (auto& cd : cds)
+        {
+            validateResult(cd, weights, std::make_pair(weight, weight), 1);
+        }
+    }
+}
+
+TEST_F(TestCalculatorDither, TestCalculatorDitherThreeEqual)
+{
+    size_t count = 0;
+    for (double weight : weight_examples)
+    {
+        tilegen::alpha::CalculatorBase::Weights weights{weight, weight, weight, weight};
+        weights[count%weights.size()] = -20.;
+        for (auto& cd : cds)
+        {
+            validateResult(cd, weights, std::make_pair(weight, weight), 1);
+        }
+        count++;
+    }
+}
+
+TEST_F(TestCalculatorDither, TestCalculatorDitherWinnerEqual)
+{
+    size_t count = 0;
+    for (double weight : weight_examples)
+    {
+        tilegen::alpha::CalculatorBase::Weights weights{weight, weight, weight - 1, weight - 50};
+        std::shuffle(weights.begin(), weights.end(), rng);
+        for (auto& cd : cds)
+        {
+            validateResult(cd, weights, std::make_pair(weight, weight), 1);
+        }
+        count++;
+    }
+}
+
+TEST_F(TestCalculatorDither, TestCalculatorDitherRunnerUpEqual)
+{
+    size_t count = 0;
+    for (double weight : weight_examples)
+    {
+        tilegen::alpha::CalculatorBase::Weights weights{weight, weight - 1, weight - 1, weight - 50};
+        std::shuffle(weights.begin(), weights.end(), rng);
+        for (auto& cd : cds)
+        {
+            validateResult(cd, weights, std::make_pair(weight, weight-1), 1);
+        }
+        count++;
+    }
+}
+
+TEST_F(TestCalculatorDither, TestCalculatorDitherAllDifferent)
+{
+    for (double weight : weight_examples)
+    {
+        double weight_runner_up = weight - (2. / 3.);
+        tilegen::alpha::CalculatorBase::Weights weights{weight, weight_runner_up, weight - 1, weight - 50};
+        std::shuffle(weights.begin(), weights.end(), rng);
+
+        validateResult(cd1, weights, std::make_pair(weight, weight_runner_up));
+        validateResult(cd2, weights, std::make_pair(weight, weight_runner_up));
+        validateResult(cd256, weights, std::make_pair(weight, weight));
+    }
+}
+
+TEST_F(TestCalculatorDither, TestCalculatorDitherRandom)
+{
+    std::uniform_real_distribution<double> weight_gen(-10, 10);
+    for (int i = 0; i<20; i++)
```

Comment:
It's just an arbitrary number of times to run the randomised test. `TestCalculatorDither::validateResult` has its own loop, so I didn't want to make this number too high. Currently this test calls `updateAlphas` 5120 times.